astropy/io/fits/util.py

```diff
@@ -857,3 +859,22 @@ def _get_array_mmap(array):
         if isinstance(base.base, mmap.mmap):
             return base.base
         base = base.base
+
+
+@contextmanager
+def _free_space_check(hdulist, dirname=None):
+    try:
+        yield
+    except OSError as exc:
+        if not isinstance(hdulist, list):
+            hdulist = [hdulist, ]
+        if dirname is None:
+            dirname = os.path.dirname(hdulist._file.name)
+        if os.path.isdir(dirname):
+            free_space = data.get_free_space_in_dir(dirname)
+            hdulist_size = np.sum(hdu.size for hdu in hdulist)
+            for hdu in hdulist:
+                hdu._close()
+            if free_space < hdulist_size:
+                raise OSError("Not enough space on disk. " + str(exc))
```

Comment:
Since you already have the size info here (how much available vs how much needed), why not include them in the error message? I would find that useful personally.